### PR TITLE
[REF] Make KFAC computation side effect free

### PR DESCRIPTION
**Pull request by f-dangel**
Before this PR, we call `.backward()` on a tensor to trigger the backward hooks that compute the grad-output-based Kronecker factors of KFAC. 
As a side effect, this will compute `.grad` of all leafs of the tensor we are differentiating.
This is problematic for use cases in optimization that use the `.grad` attribute, which is modified though by creating a `KFACLinearOperator`. 
This PR changes the `.backward()` calls into `autograd.grad(...)` which does not modify any `.grad` attributes and therefore makes the computation of KFAC side effect free.

### CODE REVIEW

File: curvlinops/kfac.py [hunk 1]
```diff
@@ -536,6 +537,10 @@ def _compute_loss_and_backward(self, output: Tensor, y: Tensor):
             output = rearrange(output, "batch ... c -> (batch ...) c")
             y = rearrange(y, "batch ... c -> (batch ...) c")
 
+        # we will differentiate w.r.t. the model parameters to trigger the backward
+        # hooks that compute the grad-output based Kronecker factors
+        params = [p for p in self._model_func.parameters() if p.requires_grad]
```
Comment (runame) [inline curvlinops/kfac.py:542]: What is the reason not to use `self._params` here?

**Review by runame — APPROVED** [review]
Looks good! Maybe you could add a unit test to avoid regressing to the same behaviour, i.e. simply check that the KFAC computation does not modify the `.grad` attribute of the model parameters.

File: curvlinops/kfac.py (hunk 1 above)
Comment (f-dangel) [reply curvlinops/kfac.py:542]: Yap, that's much better!

**Comment by f-dangel** [comment]
> Looks good! Maybe you could add a unit test to avoid regressing to the same behaviour, i.e. simply check that the KFAC computation does not modify the `.grad` attribute of the model parameters.

That was a good idea. I found one more line that needed to be removed because we were setting the gradients to zero after KFAC's backward passes.

**Comment by f-dangel** [comment]
Please take a quick second look.

**Review by runame — APPROVED** [review]
Perfect, all good now, will merge.
